pkg/controller/registry/grpc/source.go

```diff
@@ -0,0 +1,214 @@
+package grpc
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/connectivity"
+
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/controller/registry/resolver"
+	"github.com/operator-framework/operator-registry/pkg/api"
+	"github.com/operator-framework/operator-registry/pkg/api/grpc_health_v1"
+	"github.com/operator-framework/operator-registry/pkg/client"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type SourceMeta struct {
+	Address         string
+	LastConnect     metav1.Time
+	ConnectionState connectivity.State
+}
+
+type SourceState struct {
+	Key   resolver.CatalogKey
+	State connectivity.State
+}
+
+type SourceConn struct {
+	SourceMeta
+	Conn   *grpc.ClientConn
+	cancel context.CancelFunc
+}
+
+type SourceStore struct {
+	sources      map[resolver.CatalogKey]SourceConn
+	sourcesLock  sync.RWMutex
+	syncFn       func(SourceState)
+	logger       *logrus.Logger
+	notify       chan SourceState
+	timeout      time.Duration
+	readyTimeout time.Duration
+}
+
+func NewSourceStore(logger *logrus.Logger, timeout, readyTimeout time.Duration, sync func(SourceState)) *SourceStore {
+	return &SourceStore{
+		sources:      make(map[resolver.CatalogKey]SourceConn),
+		notify:       make(chan SourceState),
+		syncFn:       sync,
+		logger:       logger,
+		timeout:      timeout,
+		readyTimeout: readyTimeout,
+	}
+}
+
+func (s *SourceStore) Start(ctx context.Context) {
+	s.logger.Debug("starting source manager")
+	go func() {
```

Comment:
nit: maybe use `sync.Once` to ensure the source manager goroutine runs as a singleton.

Reply:
Seems reasonable - but I don't think we do that for our other long-running goroutines do we?

Reply:
I added that to the refactored `queueinformer.Operator` code as a safety net against successive calls that start its informers. Not saying that we _need_ to add this here - just wanted to have some feedback about whether the idea was pragmatic.